runtime/include/sandbox_set_as_running_user.h

```diff
@@ -42,9 +42,11 @@ sandbox_set_as_running_user(struct sandbox *sandbox, sandbox_state_t last_state)
 	runtime_sandbox_total_increment(SANDBOX_RUNNING_USER);
 	runtime_sandbox_total_decrement(last_state);
 
-	/* WARNING: This state change needs to be at the end of this transition because all code below this assignment
-	 * is preemptable */
+	/* WARNING: All code below this assignment is preemptable */
 	sandbox->state = SANDBOX_RUNNING_USER;
+
+	/* Now that we are preemptable, we can replay deferred sigalrms */
+	if (software_interrupt_deferred_sigalrm > 0) raise(SIGALRM);
```

Comment:
Don't you need to modify `software_interrupt_deferred_sigalrm` to be one fewer for each signal you raise? Atomically?

Do we need to worry about if SIGUSR1 is the signal to send instead?

Reply:
I don't really see value in queuing the SIGALRMs. I assume that if `software_interrupt_deferred_sigalrm` is 1..INF, we should only trigger at most a single replay. I figure the main reason that this isn't a boolean is so we can track the max number of deferred sigalrms for debugging. Still need to think through the SIGUSR1 path a bit.

Reply:
I went through the SIGUSR1 path when a cooperative sched executes the scheduler outside of a signal, and the scheduler wants to execute a sandbox in the PREEMPTED state, and it seems to be fine. The worker updates its current_sandbox to the preempted sandbox, but the sandbox doesn't transition to RUNNING_USER until inside of the SIGUSR1 signal handler (when signals are masked), so it doesn't seem like there is any point where something is unintentionally preemptable.

Reply:
I cannot follow this logic, and it seems more complicated than resetting the value before triggering deferred signals. Complexity is the enemy for this type of code.

If multiple sigalarms are queued, isn't that a bug? "Tracking for debugging" only makes sense if it shouldn't be a panicable behavior.

Reply:
An contrived example of "multiple queued sigalrms" might be a long-running WASI syscall that takes let's say 20ms running on the worker that is receiving the kernel SIGALRMs. The syscall is interrupted and receives at least three SIGALRMs, each of which increments `software_interrupt_deferred_sigalrm` by 1. When preemption is finally restored, the value of `software_interrupt_deferred_sigalrm` is 3. When the deferred sigalrms are replayed, `software_interrupt_deferred_sigalrm` is set to 0 and a single SIGALRM is raised. If we treat "multiple queued sigalrms" as a panicable behavior, that seems to imply that we need to be able to make hard guarantees that a worker will never execute more than 10ms (2x quantum) of control flow in a non-preemptable state.

Reply:
You're absolutely correct. Thanks for walking me through it!

Reply:
For the other point of "why doesn't replay clear the deferred sigalrms?," basically I clear the deferred sigalrms in two places currently: `scheduler_cooperative_sched` and `scheduler_preemptive_sched`. The reason for this is that our runtime might be calling the scheduler cooperatively while in a non-preemptive state. This is what happens when a worker is idling in what we call the `base_context`. In this state, a worker's `current_sandbox` is NULL, which means that the worker is non-preemptable. However, the base context just repeatedly calls `scheduler_cooperative_sched` over and over. This means that the scheduling logic is getting run. However, the idle loop does still respond to SIGALRMs, and the handler would still call `defer_sigalrm` when when interrupting the cooperative scheduler. Because of this, all invocations of cooperative scheduler end up zeroing this value back out. This sort of seems like a hack.

After talking this though, I think that the SIGALRM handler should perhaps only call  `defer_sigalrm` when current_sandbox is not NULL and in a non-preemptive state. If `current_sandbox` is NULL, that signifies that the worker thread is already in the scheduler in some form, so we can ignore the SIGALRM. Once we do that, we can running `software_interrupt_deferred_sigalrm_clear` in `scheduler_cooperative_sched`. That allows us to move the logic of  `software_interrupt_deferred_sigalrm_clear` into `software_interrupt_deferred_sigalrm_replay`.

Assuming that deferred_sigalrm stays a counter, rather than a bool. That ends up looking something like:

```c
static inline void
software_interrupt_counts_deferred_sigalrm_max_update(int deferred_sigalrm_count)
{
#ifdef LOG_SOFTWARE_INTERRUPT_COUNTS
	if (unlikely(deferred_sigalrm_count > software_interrupt_counts_deferred_sigalrm_max[worker_thread_idx])) {
		software_interrupt_counts_deferred_sigalrm_max[worker_thread_idx] = deferred_sigalrm_count;
	}
#endif
}

static inline void
software_interrupt_deferred_sigalrm_clear()
{
	software_interrupt_counts_deferred_sigalrm_max_update(deferred_sigalrm);
	atomic_store(&deferred_sigalrm, 0);
}

static inline void
software_interrupt_deferred_sigalrm_replay()
{
	if (deferred_sigalrm > 0) {
		software_interrupt_deferred_sigalrm_clear();
		software_interrupt_counts_deferred_sigalrm_replay_increment();
		raise(SIGALRM);
	}
}
```

There is a bunch of overhead due to the debugging stuff. If deferred sigalrm was treated as a bool, we could for sure just have a compare and swap that conditionally triggers a RAISE. Maybe we could do that if `LOG_SOFTWARE_INTERRUPT_COUNTS` is not set since we wouldn't be using the count in that case.